### PR TITLE
bootctl: clean up handling of files with no version information

### PR DESCRIPTION
**Pull request by keszybz**
get_file_version() would return:
- various negative errors if the file could not be accessed or if it was not a regular file
- 0/NULL if the file was too small
- -ESRCH or -EINVAL if the file did not contain the marker
- -ENOMEM or permissions errors
-  1 if the marker was found

bootctl status iterates over /EFI/{systemd,BOOT}/*.efi and checks if the files contain a systemd-boot version tag. Resource or permission errors should be fatal, but lack of version information should be silently ignored.

OTOH, when updating or installing bootloader files, the version is expected to be present.

get_file_version() is changed to return -ESRCH if the version is unavailable, and other errnos for permission or resource errors.

The logging is reworked to always display an error if encountered, but also to log the status at debug level what the result of the version inquiry is. This makes it figure out what is going on:
  /efi/EFI/systemd/systemd-bootx64.efi: EFI binary LoaderInfo marker: "systemd-boot 253-6.fc38"
  /efi/EFI/BOOT/BOOTfbx64.efi: EFI binary has no LoaderInfo marker.
  /efi/EFI/BOOT/BOOTIA32.EFI: EFI binary has no LoaderInfo marker.
  /efi/EFI/BOOT/BOOTX64.EFI: EFI binary LoaderInfo marker: "systemd-boot 253-6.fc38"

Replaces #27034.
Fixes https://github.com/NixOS/nixpkgs/issues/223579.

### CODE REVIEW

**Comment by keszybz** [comment]
@lnussel does this work for you?

**Comment by lnussel** [comment]
didn't review the code in depth but AFAICT lgtm

**Review by bluca — APPROVED** [review]

**Comment by Zahrun** [comment]
Thank you for this, it looks good.
I won't test it right now since NixOS uses systemd-stable v253.1 at the moment and this commit is not easily cherry-picked since it uses LOG_SET_PREFIX.
I will test this when it makes its way to systemd-stable and the repo catches up.

**Comment by bluca** [comment]
LOG_SET_PREFIX is just for logging, you can simply remove it in the backport, it will have no functional impact

**Comment by Zahrun** [comment]
I can confirm, with that line removed, the issue is fixed for me.
Is there an easy way to enable log_debug output at runtime?

**Comment by ElvishJerricco** [comment]
@bluca it'd be great if this could be backported to stable.

File: src/boot/bootctl-status.c [hunk 1]
```diff
@@ -216,15 +216,23 @@ static int enumerate_binaries(
...
+                if (!filename)
+                        return log_oom();
+                LOG_SET_PREFIX(filename);
+
                 fd = openat(dirfd(d), de->d_name, O_RDONLY|O_CLOEXEC);
                 if (fd < 0)
-                        return log_error_errno(errno, "Failed to open \"%s/%s\" for reading: %m", p, de->d_name);
+                        return log_error_errno(errno, "Failed to open file for reading: %m");
 
                 r = get_file_version(fd, &v);
```
Comment (arianvp) [inline src/boot/bootctl-status.c:228]: I noticed `bootctl status` stopped showing version numbers and I bisected it to this PR.

This PR changes the success exit code of `get_file_version()` from `1` to `0` the print logic below here hasn't been updated to reflect that.  (It only prints the version number if `r > 0`; which is now never true.

This causes `bootctl status` to not print the detected version number anymore. Seems like an unintended bug.  Is my interpretation correct?

```c

                if (r > 0)
                        r = asprintf(previous, "/%s/%s (%s%s%s)", path, de->d_name, ansi_highlight(), v, ansi_normal());
                else
                        r = asprintf(previous, "/%s/%s", path, de->d_name);
                if (r < 0)
                        return log_oom();
``` 

should simply be (given we already know that `r >= 0`

```c
                if (r == 0)
                        r = asprintf(previous, "/%s/%s (%s%s%s)", path, de->d_name, ansi_highlight(), v, ansi_normal());
```

If you agree I can send a patch to fix this.

**Pull request merged**